crates/gen_lsp_server/src/stdio.rs

```diff
@@ -37,11 +37,9 @@ pub struct Threads {
 }
 
 impl Threads {
-    pub fn join(self) -> Result<()> {
-        match self.reader.join() {
-            Ok(r) => r?,
-            Err(_) => bail!("reader panicked"),
-        }
+    pub fn exit(self) -> Result<()> {
+        // We can't rely on stdin being closed
+        drop(self.reader);
```

Comment:
Interesting!

I am pretty strongly against leaking the threads, and for [structured concurrency](https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/): always cleaning up after yourself helps with tests reproducibility and general sanity. 

However, we are hitting a "there's no way to exit a blocking read in UNIX sans sending yourself SIGUSR1" problem, because the other party may not close the pipe.

What about this hack: in `stdio_transport` we violate layering and break of the reading loop if the received message is notification with type "exit"?

Reply:
:+1:

I did look into the state of cross-platform async file I/O, but (a) it doesn't look good (b) it's a bit over my head and (c) for this small detail, a hack looks fine :)